onnxruntime/core/optimizer/computation_reduction.cc

```diff
@@ -0,0 +1,271 @@
+  return SwapGatherNDWithTargetNode(graph, gathernd_node, target_node, 0);
+}
+
+/*
+  This handler change the graphs this way:
+    Before:
+      input_1[b,s,h]   weight_2[h]
+                  \         /
+                    Add[b,s,h]    indices[b,p_s,1]
+                      |            /
+                  GatherND[b,p_s,h]
+                      |
+                <subsquent graphs>
+    After :
+               input_1[b,s,h]      indices[b,p_s,1]
+                       |            /
+                    GatherND[b,p_s,h]    weight_2[h]
+                              \              /
+                              Add[b,p_s,h]
+                                    |
+                            <subsquent graphs>
+  Note: b: batch, s: sequence_length, h: hidden_size, p_s: dynamic_prediction_count
+*/
+static Status BinaryElementwiseHandler(Graph& graph, Node& gathernd_node, Node& target_node) {
+  int target_node_input_index = GetValidInputForGatherND(target_node);
+  ORT_RETURN_IF_NOT(target_node_input_index != -1);
+  return SwapGatherNDWithTargetNode(graph, gathernd_node, target_node, target_node_input_index);
+}
+
+/*
+  This handler change the graphs this way:
+    Before:
+      input_1[b,s,h]    weight_2[h, 2h]
+                  \         /
+                  Gemm[b,s,2h]    indices[b,p_s,1]
+                      |            /
+                  GatherND[b,p_s,2h]
+                      |
+                <subsquent graphs>
+    After :
+               input_1[b,s,h]      indices[b,p_s,1]
+                       |            /
+                    GatherND[b,p_s,h]    weight_2[h,2h]
+                              \              /
+                              Gemm[b,p_s,2h]
+                                    |
+                            <subsquent graphs>
+  Note: b: batch, s: sequence_length, h: hidden_size, p_s: dynamic_prediction_count
+*/
+static Status GemmHandler(Graph& graph, Node& gathernd_node, Node& target_node) {
+  int target_node_input_index = GetValidInputForGatherND(target_node);
+  ORT_RETURN_IF_NOT(target_node_input_index == 0);
+  return SwapGatherNDWithTargetNode(graph, gathernd_node, target_node, target_node_input_index);
+}
+
+static std::unordered_map<std::string, Handler> handlers = {
+    {"Add", BinaryElementwiseHandler},
+    {"Div", BinaryElementwiseHandler},
+    {"Gelu", SimpleHandler},
+    {"Gemm", GemmHandler},
+    {"LayerNormalization", SimpleHandler},
+    {"MatMul", GemmHandler}};
+
+static Status Delegate(std::string op_type, Graph& graph, Node& gathernd_node, Node& target_node) {
+  if (handlers.count(op_type)) {
+    return handlers[op_type](graph, gathernd_node, target_node);
+  } else {
+    return common::Status(common::ONNXRUNTIME, common::NOT_IMPLEMENTED, op_type + " handler is not implemented");
+  }
+}
+
+Status ComputationReductionTransformer::ApplyImpl(Graph& graph, bool& modified, int graph_level,
```

Comment:
>ComputationReductionTransformer [](start = 7, length = 31)

Can we rename all `computation_reduction` to `gather_advance`? The current term sounds too general.

Reply:
Actually I want to leave some space for ops like Gather, GatherElements, Slice, Split, etc if needed.